File: 1836-Remove Duplicates From an Unsorted Linked List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def deleteDuplicatesUnsorted(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(float('-inf'))
        dummy.next = head
        tmp = head
        count = collections.defaultdict(int)
        
        while tmp:
            count[tmp.val] += 1
            tmp = tmp.next
        prev = dummy
        
        while prev.next:
            if count[prev.next.val] >= 2:
                prev.next = prev.next.next
            else:
                prev = prev.next
                
        return dummy.next

File: C++/1836-Remove Duplicates From an Unsorted Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicatesUnsorted(ListNode* head) {
        unordered_map<int, int> cnt;
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        ListNode* temp = head;

        while (temp) {
            cnt[temp->val]++;
            temp = temp->next;
        }

        ListNode* running = dummy;
        while (running) {
            if (running->next != nullptr && cnt[running->next->val] > 1) {
                running->next = running->next->next;
            } else {
                running = running->next;
            }
        }

        return dummy->next;
    }
};